### PR TITLE
feat: support migrating from longhorn to rook ceph

### PR DESCRIPTION
**Pull request by ricardomaraschini**
#### What this PR does / why we need it:

Enables migration from Longhorn to Rook.

#### Which issue(s) this PR fixes:
Fixes #58510

### CODE REVIEW

**Review by camilamacedo86 — APPROVED** [review]
That seems great 🥇

File: addons/rook/template/base/install.sh [hunk 1]
```diff
@@ -687,3 +692,84 @@ function rook_mds_daemons_oktostop() {
...
+
+# shows a prompt asking users for confirmation before starting to migrate data from Longhorn.
+function rook_prompt_migrate_from_longhorn() {
+    # skip on new install or when Longhorn is specified in the kURL spec
+    if [ -z "$CURRENT_KUBERNETES_VERSION" ] || [ -n "$LONGHORN_VERSION" ]; then
+        return 0
+    fi
+
+    # do not proceed if Longhorn is not installed
+    if ! kubectl get ns | grep -q longhorn-system; then
```
Comment (emosbaugh) [inline addons/rook/template/base/install.sh:753]: does pvmigrate delete the namespace?

File: addons/rook/template/base/install.sh [hunk 2]
```diff
@@ -687,3 +692,84 @@ function rook_mds_daemons_oktostop() {
...
+    do
+        # run validation checks for non default Longhorn storage classes
+        if ! longhorn_scs_pvmigrate_dryrun_output=$($BIN_PVMIGRATE --source-sc "$longhorn_sc" --dest-sc "$rook_storage_class" --rsync-image "$KURL_UTIL_IMAGE" --preflight-validation-only) ; then
+            break
+        fi
+    done
+
+    if [ -n "$longhorn_default_sc" ] ; then
+        # run validation checks for Rook default storage class
+        longhorn_default_sc_pvmigrate_dryrun_output=$($BIN_PVMIGRATE --source-sc "$longhorn_default_sc" --dest-sc "$rook_storage_class" --rsync-image "$KURL_UTIL_IMAGE" --preflight-validation-only || true)
```
Comment (emosbaugh) [inline addons/rook/template/base/install.sh:733]: and this

```suggestion
        if longhorn_default_sc_pvmigrate_dryrun_output=$($BIN_PVMIGRATE --source-sc "$longhorn_default_sc" --dest-sc "$rook_storage_class" --rsync-image "$KURL_UTIL_IMAGE" --preflight-validation-only) ; then
            longhorn_default_sc_pvmigrate_dryrun_output=""
        fi
```

File: addons/rook/template/base/install.sh [hunk 3]
```diff
@@ -104,6 +106,9 @@ function rook() {
     if ! spinner_until 120 rook_rgw_is_healthy ; then
         bail "Failed to detect healthy rook-ceph object store"
     fi
+
+    # migrate from Longhorn storage if applicable
+    rook_maybe_migrate_from_longhorn
```
[the remark this answers is not in the document]
Comment (ricardomaraschini) [reply addons/rook/template/base/install.sh:111]: Pardon, I could not understand your question. The function `rook_maybe_migrate_from_longhorn` is called as part of the `rook()` function (not sure about what `post_init` you are referring here). Can you elaborate ?

File: addons/rook/template/base/install.sh (hunk 1 above)
Comment (ricardomaraschini) [reply addons/rook/template/base/install.sh:753]: No, the namespace is deleted [here](https://github.com/replicatedhq/kURL/blob/main/scripts/install.sh#L561), for sake of simplicity [this](https://github.com/replicatedhq/kURL/blob/main/scripts/common/longhorn.sh#L129-L138) is the function body. We set `DID_MIGRATE_LONGHORN_PVCS` [here](https://github.com/replicatedhq/kURL/pull/3919/files#diff-ac769affc747039b8657a32acfaf395c37caaa34ff469b587220806ff576655bR706).

File: addons/rook/template/base/install.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (emosbaugh) [reply addons/rook/template/base/install.sh:111]: my mistake. the github ui was confusing

**Review by emosbaugh — APPROVED** [review]

**Pull request merged**